### ISSUE TITLE
SmtpClient is slow when Exchange server has tarpit enabled

### ISSUE THREAD

**Issue by PrzemyslawKlys** [jstedfast/MailKit#1314]
I have a weird issue with one of my clients. I'm using [Mailozaurr](https://github.com/EvotecIT/Mailozaurr) to send emails (based on your library). 

1. When I send an email using Send-MailMessage (builtin 5.1 PowerShell) it takes 0.5-0.6 seconds to send an email thru 4x Exchange 2016 servers, and also the same time using Load Balancer which is in front of it. 
2. When I send an email using Send-EmailMessage (from Mailozaurr) it takes 0.7 seconds to send an email thru 4x Exchange 2016, but 5 to 10 seconds using Load Balancer. 

It behaves the same whether TLS is on or off. How would I verify what could be the reason Mailkit triggers a problem with Load Balancer? I'm pretty sure it's nothing in your library - more like my code, or more likely LB doing something fancy with responding that takes too long on timeout or something.

**Comment by jstedfast** [comment]
One possibility might be to enable timestamps in MailKit's protocol logger. That won't help with connection performance (which could be an issue, I suppose?), but it might give us an idea of where the delays are in terms of SMTP commands once the connection is established.

**Comment by PrzemyslawKlys** [comment]
Thank you. I've added logging capabilities to my module and it seems the delay happens after the second EHLO. 

```
2022-01-07 08:34:41:280 Connected to smtp://smtp-i.domain.local:587/?starttls=always
2022-01-07 08:34:41:281 S: 220 smtp-i.domain.local ESMTP mail service ready
2022-01-07 08:34:41:281 C: EHLO [172.16.11.10]
2022-01-07 08:34:41:323 S: 250-smtp-i.domain.local Hello [172.16.11.10]
2022-01-07 08:34:41:323 S: 250-SIZE 32505856
2022-01-07 08:34:41:323 S: 250-DSN
2022-01-07 08:34:41:323 S: 250-ENHANCEDSTATUSCODES
2022-01-07 08:34:41:323 S: 250-AUTH GSSAPI NTLM LOGIN
2022-01-07 08:34:41:323 S: 250-8BITMIME
2022-01-07 08:34:41:323 S: 250-BINARYMIME
2022-01-07 08:34:41:323 S: 250-STARTTLS
2022-01-07 08:34:41:323 S: 250 CHUNKING
2022-01-07 08:34:41:323 C: STARTTLS
2022-01-07 08:34:41:324 S: 220 Go ahead
2022-01-07 08:34:41:361 C: EHLO [172.16.11.10]
2022-01-07 08:34:46:373 S: 250-smtp-i.domain.local Hello [172.16.11.10]
2022-01-07 08:34:46:374 S: 250-SIZE 32505856
2022-01-07 08:34:46:374 S: 250-DSN
2022-01-07 08:34:46:374 S: 250-ENHANCEDSTATUSCODES
2022-01-07 08:34:46:374 S: 250-AUTH GSSAPI NTLM LOGIN
2022-01-07 08:34:46:374 S: 250-8BITMIME
2022-01-07 08:34:46:374 S: 250-BINARYMIME
2022-01-07 08:34:46:374 S: 250 CHUNKING
2022-01-07 08:34:46:375 C: AUTH LOGIN
2022-01-07 08:34:46:376 S: 334 VXNlcm5hbWU6
2022-01-07 08:34:46:376 C: ********
2022-01-07 08:34:46:379 S: 334 UGFzc3dvcmQ6
2022-01-07 08:34:46:379 C: ********
2022-01-07 08:34:46:502 S: 235 2.7.0 Authentication successful
2022-01-07 08:34:46:502 C: MAIL FROM:<serviceaccount@testdomain.com> SIZE=783
2022-01-07 08:34:46:502 S: 250 2.1.0 Sender OK
2022-01-07 08:34:46:502 C: RCPT TO:<przemyslawklys@testdomain.com>
2022-01-07 08:34:46:502 S: 250 2.1.5 Recipient OK
2022-01-07 08:34:46:502 C: DATA
2022-01-07 08:34:46:516 S: 354 Start mail input; end with <CRLF>.<CRLF>
2022-01-07 08:34:46:517 C: From: serviceaccount@testdomain.com
2022-01-07 08:34:46:517 C: Date: Fri, 07 Jan 2022 09:34:41 +0100
2022-01-07 08:34:46:517 C: Subject: Test Using smtp-i.domain.local, with SSL True - trying 0
2022-01-07 08:34:46:517 C: Message-Id: <EKD84UY3RFU4.MZNV2JFLN4XK3@SendingClient>
2022-01-07 08:34:46:517 C: To: przemyslawklys@testdomain.com
2022-01-07 08:34:46:517 C: MIME-Version: 1.0
2022-01-07 08:34:46:517 C: Content-Type: text/html; charset=utf-8
2022-01-07 08:34:46:517 C: 
2022-01-07 08:34:46:517 C: <!-- saved from url=(0014)about:internet -->
2022-01-07 08:34:46:517 C:  <!DOCTYPE html>
2022-01-07 08:34:46:517 C:  <html></html>
2022-01-07 08:34:46:517 C: .
2022-01-07 08:34:46:648 S: 250 2.6.0 <EKD84UY3RFU4.MZNV2JFLN4XK3@SendingClient> [InternalId=6030134101258, Hostname=SERVEREX002.domain.ad] 1889 bytes in 0.133, 13.776 KB/sec Queued mail for delivery
2022-01-07 08:34:46:649 C: QUIT
2022-01-07 08:34:46:651 S: 221 2.0.0 Service closing transmission channel
```

So it would seem a delay on Exchange Server

**Comment by jstedfast** [comment]
After the STARTTLS command, a server can offer different capabilities.

**Comment by jstedfast** [comment]
FWIW, this is the relevant section of the SMTP STARTTLS specification: https://datatracker.ietf.org/doc/html/rfc2487#section-5.2

**Comment by PrzemyslawKlys** [comment]
Can one skip the first part and start directly with starttls?

**Comment by jstedfast** [comment]
Possibly, but you'd have no way of knowing if the server supported the STARTTLS command without issuing an EHLO command first. Might be ok because you could just issue a STARTTLS command and handle it based on command errors.

I think some SMTP servers reject any commands until an EHLO command is sent, though.

That 5 second delay is killer.

**Comment by PrzemyslawKlys** [comment]
Funny thing is when we use native smtp library there is no delay 😮‍💨 and I fail to see why....

**Comment by jstedfast** [comment]
Do they send both EHLO commands?

I noticed in your log that both EHLO responses were identical (other than not listing STARTTLS in the second response).

The main difference that I would expect would be different AUTH mechanisms. For example, LOGIN and PLAIN are pretty weak (just base64), so a server might not list those pre-STARTTLS but might feel safe with them once TLS encryption is enabled.

**Comment by jstedfast** [comment]
Right now, as far as we know, the main delay is that second EHLO command response which takes ~5 seconds to receive. That's huge.

I think that skipping the first EHLO and going directly to a STARTTLS command is not likely to drastically reduce delays because we'd still end up sending that second EHLO command which is where the delay happens. It's possible that the server will respond instantly for the first EHLO, even if it comes after a STARTTLS, but I wouldn't bet on it.

I think that the best option would be to avoid that second EHLO but I feel a bit unsafe in doing that because the spec specifically states that clients SHOULD send a second EHLO after STARTTLS and that clients MUST forget every capability previously reported.

What *may* be happening is that the server is taking a while to do a reverse lookup of the IP address and for whatever reason, that is taking a while the second time. Perhaps if a hostname is used instead, it would be faster?

If System.Net.Mail uses a hostname, maybe we can find a way to do that as well. There is already code that *tries* to get a hostname, but it might be possible to improve the logic so that it doesn't fail and have to fall back to using an IP address.

**Comment by PrzemyslawKlys** [comment]
I don't know what the native library does tho. Would need to find a way to enable logger just like in your case and I haven't seen any docs for that. This would be best to compare 1 to 1 - why different behavior. 

We also had a session with the network team capturing traffic via Load Balancer and testing both options - I should have some more information regarding that once they do analysis. 

The other weird thing is - if I target directly Exchange server without LB both behave the same... but only during tests. When I run the full script which sends 800 emails it takes like 9 hours to do so, sometimes going up to 16 hours when it's supposed to send 1000. 

It's a bit tricky to know why all that is happening, as it used to work fine 2-3 months back and everyone says they changed nothing.

**Comment by PrzemyslawKlys** [comment]
I've been trying to reproduce this issue and doing things manually and it seems that if I do:

- `telnet smtp.office365.com 25`
- Then write `ehlo` it will return very fast
- If i type second ehlo it will return in 5 seconds.

I am not sure what to think about it. Either I do something wrong, or I don't know.

**Comment by PrzemyslawKlys** [comment]
So it seems Exchange have TarpitInterval Enabled for unauthenticated sending. I asked this 2 months ago and they said there is no tarpit. Today I've gave them command and told them to check and sure enough Tarpit is enabled. 

However, the Tarpit is only used in specific cases: 

> The amount of time to artificially delay SMTP responses to unauthenticated remote servers that appear to be abusing the connection. Authenticated connections are never delayed in this manner.

So it would seem that the native library somehow "skips" this step and goes thru without being limited. The question is why and how could we get Mailkit to do the same thing ?

**Comment by jstedfast** [comment]
This is where I wish I had logs from the other SmtpClient implementation. Is it skipping the re-EHLO? If it does, can it really rely on the initial SMTP capabilities? The problem is that these capabilities *can* change (especially the available AUTH mechanisms), so it seems dangerous to not send EHLO again after STARTTLS.

**Comment by jstedfast** [comment]
I guess it *must* be skipping the re-EHLO. I don't think there's any other explanation.

**Comment by PrzemyslawKlys** [comment]
You know we are talking about https://docs.microsoft.com/en-us/dotnet/api/system.net.mail.smtpclient?view=net-6.0 because that's what I use in the 2nd library right? You didn't ever did comparison on how both works while working on Mailkit?

**Comment by PrzemyslawKlys** [comment]
But maybe it does some NTLM authentication? Something that skips tarpit altogether by authenticating before it even gets to the point of tarpit?

**Comment by jstedfast** [comment]
No, it only authenticates after STARTTLS afaict.

Keep in mind, too, that System.Net.Mail.SmtpClient re-uses past SMTP connections via a connection pool which means they don't have to go through the connect/starttls/auth process each time you create an SmtpClient and send a message.

**Comment by jstedfast** [comment]
If you set the `SmtpClient.LocalDomain` property to your system's host name, does that improve perf?

**Comment by jstedfast** [comment]
(or just hard-code "localhost" for now, just to check things)

**Comment by jstedfast** [comment]
The above commit essentially makes use of a default LocalDomain value.

MailKit has historically used an IP address in the EHLO/HELO that it got from the `Socket`'s `LocalEndPoint` if no `LocalDomain` property was set. The above change obtains the system's hostname to use in the EHLO/HELO instead.

The reason for this change is that it *might* slow things down if the SMTP server tries to do a reverse lookup of the IP address to get a proper hostname (which I think most do). That said, I'm not sure using the local machine's hostname is necessarily any better if the SMTP server decides to run a DNS query on it because, at least in my case, DNS isn't going to find anything.

I had a patch that did the DNS query client-side and cached the result for later EHLO commands (e.g. for after STARTTLS and/or after AUTH if AUTH negotiates a security layer), but gave up on that since it was more complicated and because performing that step was more likely to introduce more performance problems than it solved.

I think the real reason that System.Net.Mail is faster is more likely because of connection pooling (see my comment from 3 days ago) and not anything else.

**Comment by PrzemyslawKlys** [comment]
I'm still going to test two things:
1. I've added to mailozaurr ability to use localhost or any predefined value there - so I'll test localhost + hostname
2. I've added logging which I will test on a global scale

When we disabled tarpit - it worked ok when running manually via service account, but still when running with MSA account (with service account authorization) the delay was from 5 seconds to even 1minute long. 

This is getting exhausting because it's hard to be 100% sure what is the problem and where to look for it. So I'm going to add some more logging to the production script and randomly save transcription of communication to see what other reason it may be. 

I don't think it's the real reason with connection polling because on 13th of October it worked fine sending 600 emails with less than 1 hour. After 13th - 15h hours to do send the same. Something changed and I'm trying to find out what. They were exchanging old Exchange 2010 to Exchange 2016 servers sometime in 2021, maybe DNS changes. 

I appreciate that you're trying to optimize it even more - but all in all - the delay is way too big to be something in your code. I could add parallel processing of SMTP sending - but I'm pretty sure it's something that's just on company side - the question is what.

Btw. can it be made so that connection polling will be ever available in Mailkit (for further performance enhancements?)

**Comment by jstedfast** [comment]
> Btw. can it be made so that connection polling will be ever available in Mailkit (for further performance enhancements?)

Probably not because the API would make that difficult.

The way this works in System.Net.Mail is that the SmtpClient hides all of the connect/authentication logic and just exposes a Send() method which will either grab an existing connection from the pool or else create a new connection (and authenticate).

MailKit's API has direct Connect() and Authenticate() methods.

**Comment by PrzemyslawKlys** [comment]
So, I've found a real reason behind this. I'm sending emails to users. My function which creates an email if you don't give any path to it - creates a random temporary path (or so I thought) using `[System.IO.Path]::GetTempFileName()` - this is so the HTML can always be saved if the user decides to show it up after creation.

The thing is - that method actually creates a file, not just creates a string with a file name to be used. And while i was testing i  was using my standard id, but the script was running on MSA account. 

Right now the script is sending 1 email per 40 seconds - i was getting furious at the Exchange team. Until I've opened with PSexec PowerShell window on that service account and opening simple HTML with my module was taking a very long time, and it started throwing errors on that method `[System.IO.Path]::GetTempFileName()` that it couldn't get it done.

Then I go into temp folder, list files and after 1 minute of listing I now understand - there are millions of files there, so creating a new file with "temporary" name is taking so long it actually impacts sending emails (which as part of the sending - create HTML using that method within)

![image](https://user-images.githubusercontent.com/15063294/150761461-244cd8d1-4812-4f57-9013-608af129da06.png)

After 10 minutes waiting - delete command was done :/ 

I guess I should have read what that method does actually - i always thought it's just creating random name that I can use - not create a file on the file system :/
